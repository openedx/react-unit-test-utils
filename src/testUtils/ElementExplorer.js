import { isEqual } from 'lodash';

class ElementExplorer {
  constructor(element, toSnapshot, parent = null) {
    this.el = element;
    this.toSnapshot = toSnapshot;
    this.props = {};
    this.type = null;
    this.children = [];
    this.parent = parent;
    if (element.type) {
      this.props = element.props;
      this.type = element.type;
    }
    if (element.children) {
      this.children = element.children.map(child => new ElementExplorer(child, toSnapshot, this));
    }
  }

  findByTestId(testId) {
    const elements = [];
    const findChildrenByTestId = (el) => {
      if (el.props['data-testid'] === testId) {
        elements.push(el);
      }
      el.children.forEach(child => {
        findChildrenByTestId(child);
      });
    };
    findChildrenByTestId(this);
    return elements;
  }

  findByType(type) {
    const typeString = typeof type === 'string' ? type : type.name;
    const elements = [];
    const findChildrenByType = (el) => {
      if (el.type === typeString) {
        elements.push(el);
      }
      el.children.forEach(child => {
        findChildrenByType(child);
      });
    };
    findChildrenByType(this);
    return elements;
  }

  get data() {
    const out = {
      props: this.props,
      type: this.type,
      children: this.children.map(child => child.data),
    };
    if (this.type === null) {
      out.el = this.el;
    }
    return out;
  }

  matches(el) {
    const toCheck = el.data ? el.data : el;
    return isEqual(toCheck, this.data);
  }

  get snapshot() {
    return this.toSnapshot(this.el);
  }
}

export default ElementExplorer;
