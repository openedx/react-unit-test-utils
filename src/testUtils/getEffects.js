import { isEqual } from 'lodash';
/**
 * getEffect(prereqs, reactModule)
 * Returns the callbacks passed to useEffect mock call with the given prerequisites key,
 * or null if none match.
 *
 * @param {any[]} prereqs - unique prerequisite list identifying the target useEffect call
 * @param {object} reactModule - react module to be mocked.
 * @return {function} - callback method passed for the effect hook.
 */
const getEffects = (prereqs, reactModule) => {
  const { calls } = reactModule.useEffect.mock;
  const match = calls.filter(call => isEqual(call[1], prereqs));
  return match.length ? match[0] : null;
};

export default getEffects;
