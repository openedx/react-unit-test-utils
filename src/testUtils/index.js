export { default as formatMessage } from './formatMessage';
export { default as getEffects } from './getEffects';
export { default as mockComponents } from './mockComponents';
export { default as mockUseKeyedState } from './mockUseKeyedState';
export { default as shallow } from './shallow';
