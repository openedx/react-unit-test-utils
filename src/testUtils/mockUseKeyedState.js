import * as utils from '../utils';

/**
 * Mock utility for working with useState in a hooks module.
 * Expects/requires an object containing the state object in order to ensure
 * the mock behavior works appropriately.
 *
 * Expected format:
 *   hooks = { state: { <key>: (val) => React.createRef(val), ... } }
 *
 * Returns a utility for mocking useState and providing access to specific state values
 * and setState methods, as well as allowing per-test configuration of useState value returns.
 *
 * Example usage:
 *   // hooks.js
 *   import * as module from './hooks';
 *   import { useKeyedState, StrictDict } from '@edx/react-unit-test-utils';
 *
 *   const stateKeys = StrictDict({
 *     isOpen: 'isOpen',
 *     hasDoors: 'hasDoors',
 *     selected: 'selected',
 *   });
 *   ...
 *   export const exampleHook = () => {
 *     const [isOpen, setIsOpen] = useKeyedState(stateKeys.isOpen, false);
 *     const handleClickOpen = () => {
 *       setIsOpen(true);
 *     };
 *     if (!isOpen) { return null; }
 *     return { isOpen, setIsOpen, handleClickOpen };
 *   }
 *   ...
 *
 *   // hooks.test.js
 *   import * as hooks from './hooks';
 *   import { mockUseKeyedState } from '@edx/react-unit-test-utils';
 *   const state = mockUseKeyedState(hooks.stateKeys)
 *   ...
 *   describe('exampleHook', () => {
 *     let out;
 *     beforeEach(() => {
 *       state.mock();
 *       out = exampleHook();
 *     });
 *     describe('behavior', () => {
 *       it('initializes state hooks', () => {
 *         state.expectInitializedWith(state.keys.isOpen, false);
 *       });
 *     });
 *     describe('output', () => {
 *       it('returns null if isOpen is default value', () => {
 *         expect(out).toEqual(null);
 *       });
 *       describe('if isOpen is not false', () => {
 *         beforeEach(() => {
 *           state.mockVal(state.keys.isOpen, true);
 *           out = exampleHook();
 *         });
 *         it('forwards isOpen and setIsOpen from state hook', () => {
 *           expect(out.isOpen).toEqual(state.values.isOpen);
 *           expect(out.setIsOpen).toEqual(state.setState.isOpen);
 *         });
 *         test('handleClickOpen sets isOpen to true', () => {
 *           out.handleClickOpen();
 *           state.expectSetStateCalledWith(state.keys.isOpen, true);
 *         });
 *       });
 *     });
 *     afterEach(() => { state.restore(); });
 *   });
 *
 * @param {obj} hooks - hooks module containing a 'state' object
 */
export class MockUseKeyedState {
  constructor(stateKeys) {
    this.keys = stateKeys;
    this.hookSpy = jest.spyOn(utils, 'useKeyedState');
    this.values = {};
    this.initValues = {};
    this.setState = {};

    this.mock = this.mock.bind(this);
    this.restore = this.restore.bind(this);
    this.mockVal = this.mockVal.bind(this);
    this.setup = this.setup.bind(this);
    this.mockHook = this.mockHook.bind(this);
    this.mockVals = this.mockVals.bind(this);
    this.resetVals = this.resetVals.bind(this);
    this.setup();
  }

  setup() {
    this.values = {};
    this.initValues = {};
    this.setState = {};
    Object.keys(this.keys).forEach((key) => {
      this.values[key] = null;
      this.setState[key] = jest.fn((val) => {
        this.values[key] = val;
      });
    });
  }

  mockHook(key, val) {
    this.initValues[key] = val;
    this.values[key] = val;
    return [this.values[key], this.setState[key]];
  }

  /**
   * Replace the hook module's state object with a mocked version, initialized to default values.
   */
  mock() {
    this.hookSpy.mockImplementation(this.mockHook);
  }

  /**
   * Mock the state getter associated with a single key to return a specific value one time.
   *
   * @param {string} mockKey - state key (from this.keys)
   * @param {any} val - new value to be returned by the useState call.
   */
  mockVal(mockKey, val) {
    this.hookSpy.mockImplementationOnce((key, newVal) => {
      if (key === mockKey) {
        this.values[key] = val;
        this.initValues[key] = newVal;
        return [val, this.setState[key]];
      }
      return this.mockHook(key);
    });
  }

  /**
   * Mock the state getters associated with set of single keys to return specific values one time.
   *
   * @param {object} mapping - { <stateKey>: <val to return> }
   */
  mockVals(mapping) {
    this.hookSpy.mockImplementation((key, val) => {
      if (mapping[key]) {
        this.values[key] = mapping[key];
        this.initValues[key] = val;
        return [mapping[key], this.setState[key]];
      }
      return this.mockHook(key);
    });
  }

  /**
   * Reset hook value mappings after multi-value override
   */
  resetVals() {
    this.hookSpy.mockImplementation(this.mockHook);
  }

  /**
   * Restore the hook module's state object to the actual code.
   */
  restore() {
    this.hookSpy.mockRestore();
  }

  expectInitializedWith(key, value) {
    expect(this.initValues[key]).toEqual(value);
  }

  expectSetStateCalledWith(key, value) {
    expect(this.setState[key]).toHaveBeenCalledWith(value);
  }
}

export const mockUseKeyedState = (stateKeys) => new MockUseKeyedState(stateKeys);

export default mockUseKeyedState;
