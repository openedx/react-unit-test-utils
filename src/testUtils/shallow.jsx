/* eslint-disable import/no-extraneous-dependencies */
import React from 'react';
import ShallowRenderer from 'react-test-renderer/shallow';
import {
  isFragment,
  isLazy,
  isPortal,
  isMemo,
  isSuspense,
  isForwardRef,
} from 'react-is';
import ElementExplorer from './ElementExplorer';

const getNodeName = node => node.displayName || node.name || '';

class ReactShallowRenderer {
  shallowRenderer = null;

  constructor(children, { Wrapper = null } = {}) {
    this.shallowRenderer = new ShallowRenderer();
    this.shallowWrapper = Wrapper
      ? this.shallowRenderer.render(<Wrapper>{children}</Wrapper>)
      : this.shallowRenderer.render(children);
  }

  static shallow(Component) {
    let out;
    try {
      out = new ReactShallowRenderer(Component);
    } catch (error) {
      console.log({ error, Component });
      out = ReactShallowRenderer.flatRender(Component);
      console.log({ out });
    }
    return out;
  }

  static flatRender(child) {
    const loadEl = (toLoad) => {
      if (typeof toLoad === 'object') {
        const { children, ...props } = toLoad.props;
        const el = { type: toLoad.type, props };
        if (Array.isArray(children)) {
          el.children = children.map(loadEl);
        } else if (children !== undefined) {
          el.children = [loadEl(children)];
        } else {
          el.children = [];
        }
        return el;
      }
      // custom return for basic jsx components (mostly for shallow comparison)
      return {
        el: toLoad,
        type: null,
        props: {},
        children: [],
      };
    };
    return loadEl(child);
  }

  isEmptyRender() {
    const data = this.getRenderOutput();
    return data === nul || data === false;
  }

  get snapshot() {
    return this.getRenderOutput(true);
  }

  get instance() {
    return new ElementExplorer(this.getRenderOutput(), ReactShallowRenderer.toSnapshot);
  }

  extractType(node) {
    if (typeof node === 'string') { return node; }

    const name = getNodeName(node.type) || node.type || 'Component';

    if (isLazy(node)) { return 'Lazy'; }

    if (isMemo(node)) { return `Memo(${name || this.extractType(node.type)})`; }

    if (isSuspense(node)) { return 'Suspense'; }

    if (isPortal(node)) { return 'Portal'; }

    if (isFragment(node)) { return 'Fragment'; }

    if (isForwardRef(node)) { return this.getWrappedName(node, node.type.render, 'ForwardRef'); }

    return name;
  }

  getWrappedName(outerNode, innerNode, wrapperName) {
    const functionName = getNodeName(innerNode);
    return outerNode.type.displayName || (functionName !== '' ? `${wrapperName}(${functionName})` : wrapperName);
  }

  getRenderOutput(render = false) {
    if (!this.shallowWrapper) {
      return this.shallowWrapper;
    }
    return this.transformNode(this.shallowWrapper, render);
  }

  // eslint-disable-next-line
  extractProps({ children, ...props } = {}, key, render) {
    const childrenArray = Array.isArray(children) ? children : [children];
    return {
      children: childrenArray.filter(Boolean).flatMap(
        (node) => this.transformNode(node, render),
      ),
      props: {
        ...props,
        ...(key ? { key } : {}),
      },
    };
  }

  transformNode(node, render) {
    if (Array.isArray(node)) {
      return node.map((n) => this.transformNode(n, render));
    }
    if (typeof node !== 'object') {
      return node;
    }
    const out = {
      type: this.extractType(node),
      ...this.extractProps(node.props, node.key, render),
    };
    if (render) {
      // this symbol is used by Jest to prettify serialized React test objects:
      // https://github.com/facebook/jest/blob/e0b33b74b5afd738edc183858b5c34053cfc26dd/packages/pretty-format/src/plugins/ReactTestComponent.ts
      out.$$typeof = Symbol.for('react.test.json');
    }
    return out;
  }

  static toSnapshot(data) {
    return {
      ...data,
      $$typeof: Symbol.for('react.test.json'),
      children: data.children.map(ReactShallowRenderer.toSnapshot),
    };
  }
}

export default ReactShallowRenderer.shallow;
