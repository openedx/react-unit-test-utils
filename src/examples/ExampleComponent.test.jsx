import React from 'react';
import { formatMessage, shallow } from 'testUtils';
import { Form } from '@edx/paragon';
import { useIntl } from '@edx/frontend-platform/i18n';

import ImportedComponent from './ImportedComponent';
import useExampleComponentData from './hooks';
import ExampleComponent, { testIds } from './ExampleComponent';
import messages from './messages';

jest.mock('./ImportedComponent', () => 'ImportedComponent');
jest.mock('./hooks', () => ({ __esModule: true, default: jest.fn() }));

let el;
let hookProps;
describe('ExampleComponent component', () => {
  beforeAll(() => {
    hookProps = {
      fileInputRef: { current: null },
      gradeExportUrl: 'test-grade-export-url',
      handleClickImportGrades: jest.fn().mockName('hooks.handleClickImportGrades'),
      handleFileInputChange: jest.fn().mockName('hooks.handleFileInputChange'),
    };
    useExampleComponentData.mockReturnValue(hookProps);
    el = shallow(<ExampleComponent />);
  });
  describe('behavior', () => {
    it('initializes hooks', () => {
      expect(useExampleComponentData).toHaveBeenCalledWith();
      expect(useIntl).toHaveBeenCalledWith();
    });
  });
  describe('render', () => {
    test('snapshot', () => {
      expect(el.snapshot).toMatchSnapshot();
    });
    describe('output', () => {
      test('Form', () => {
        const control = el.instance.findByType(Form)[0];
        expect(control.props.action).toEqual(hookProps.formAction);
      });
      test('FileControl', () => {
        const control = el.instance.findByTestId(testIds.fileControl)[0];
        expect(control.props.onChange).toEqual(hookProps.handleFileInputChange);
      });
      test('imported component', () => {
        const control = el.instance.findByType(ImportedComponent)[0];
        expect(control.props.onClick).toEqual(hookProps.handleClickImportedComponent);
      });
      test('random', () => {
        const control = el.instance.findByType('div')[0];
        expect(control.matches(shallow(
          <div>
            <h1>{formatMessage(messages.heading)}</h1>
            <span>{formatMessage(messages.span)}</span>
          </div>,
        ))).toEqual(true);
      });
    });
  });
});
