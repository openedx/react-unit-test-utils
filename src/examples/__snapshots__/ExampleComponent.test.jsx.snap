// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ExampleComponent component render snapshot 1`] = `
<Fragment>
  <Form
    method="post"
  >
    <Form.Group
      controlId="csv"
    >
      <Form.Control
        data-testid="file-control"
        label="A File Control"
        onChange={[MockFunction hooks.handleFileInputChange]}
        type="file"
      />
    </Form.Group>
  </Form>
  <ImportedComponent
    className="imported-component"
    label={
      {
        "defaultMessage": "I am imported",
        "description": "File Control Label",
        "id": "examples.importedComponentLabel",
      }
    }
  />
  <div>
    <h1>
      I am a heading
    </h1>
    <span>
      I am a span
    </span>
  </div>
</Fragment>
`;
